### PR TITLE
feat: read datastore information from config when migrating database

### PR DESCRIPTION
**Pull request by suttod**
Read datasorce from config when migrating

## Description

Datasource engine and URI can be specified via config file for `openfga run`. Adding this feature allows the same configuration file to be used for `openfga migrate` as well. Values specified in config file have lower precedence than env / cli flag (as also defined by viper).

## References

There was a [stub](https://github.com/openfga/openfga/blob/v0.4.2/cmd/migrate.go#L39) in the source code for using config file, but actually had no effect.

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

**Comment by adriantam** [comment]
Hello @suttod !  Thank you for contributing!

File: cmd/migrate.go [hunk 1]
```diff
@@ -20,7 +20,9 @@ import (
 
 const (
 	datastoreEngineFlag = "datastore-engine"
+	datastoreEngineConf = "datastore.engine"
 	datastoreURIFlag    = "datastore-uri"
+	datastoreURIConf    = "datastore-uri"
```
Comment (adriantam) [inline cmd/migrate.go:25]:
```suggestion
	datastoreURIConf    = "datastore.uri"
```

Do you mean this?

File: cmd/migrate.go [hunk 2]
```diff
@@ -41,6 +43,12 @@ func NewMigrateCommand() *cobra.Command {
 		viper.AddConfigPath(path)
 	}
 
+	err := viper.ReadInConfig()
+	if err == nil {
+		viper.SetDefault(datastoreEngineFlag, viper.Get(datastoreEngineConf))
```
Comment (adriantam) [inline cmd/migrate.go:48]: why are we setting the default after reading the config file?

Comment (suttod) [reply cmd/migrate.go:48]: When reading the config file, the values are accessible as nested keys (ie. `datastore.uri`). In `runMigration` value is accessed as `viper.GetString(datastoreURIFlag)`, that resolves to `datastore-engine` (ie. with dash).

**Review by miparnisari — APPROVED** [review]
Tested like so:

```shell
[28/03/23 5:51:53] ~/GitHub/openfga (main) $ cat /Users/mparnisari/.openfga/config
datastore:
   engine: "postgres"
   uri: "postgres://postgres:password@127.0.0.1:5432/postgres?sslmode=disable"
[28/03/23 5:51:56] ~/GitHub/openfga (main) $ make build && ./openfga migrate      
go build -o ./openfga ./cmd/openfga
Error: unknown datastore engine type: 
Usage:
  openfga migrate [flags]

Flags:
      --datastore-engine string                                                                        (required) the datastore engine that will be used for persistence
      --datastore-uri string                                                                           (required) the connection uri of the database to run the migrations against (e.g. 'postgres://postgres:password@localhost:5432/postgres')
  -h, --help                                                                                           help for migrate
      --timeout duration                                                                               a timeout after which the migration process will terminate (default 1m0s)
      --version the version to migrate to. If omitted, the latest version of the schema will be used   the version to migrate to. If omitted, the latest version of the schema will be used

[28/03/23 5:52:21] ~/GitHub/openfga (main) $ gh pr checkout 655
Switched to branch 'read-conf-file-when-migrate'
[28/03/23 5:52:25] ~/GitHub/openfga (read-conf-file-when-migrate) $ make build && ./openfga migrate
go build -o ./openfga ./cmd/openfga
[28/03/23 5:52:29] ~/GitHub/openfga (read-conf-file-when-migrate) $ 
```

Thanks for your contribution @suttod!

File: cmd/migrate.go [hunk 3]
```diff
@@ -41,6 +43,12 @@ func NewMigrateCommand() *cobra.Command {
...
+		viper.SetDefault(datastoreURIFlag, viper.Get(datastoreURIConf))
+	}
```
Comment (miparnisari) [inline cmd/migrate.go:50]:
```suggestion
	err := viper.ReadInConfig()
	if err == nil {
		if !viper.IsSet(datastoreEngineConf) {
			log.Fatalf("missing required field: %v", datastoreEngineFlag)
		}
		if !viper.IsSet(datastoreURIConf) {
			log.Fatalf("missing required field: %v", datastoreURIFlag)
		}
		viper.SetDefault(datastoreEngineFlag, viper.Get(datastoreEngineConf))
		viper.SetDefault(datastoreURIFlag, viper.Get(datastoreURIConf))
	}
```

**Review by miparnisari — CHANGES_REQUESTED** [review]

File: cmd/migrate.go (hunk 3 above)
Comment (suttod) [reply cmd/migrate.go:50]: Well it is not checked either by the `run` command. Setting the `engine` to `postgres` there without an `uri` produces the following output:

    $ cat ~/.openfga/config.yaml
    datastore:
      engine: postgres
    
    $ ./openfga run             
    2023-03-31T18:13:59.674+0200	INFO	🧪 experimental features enabled: []
    2023-03-31T18:13:59.675+0200	INFO	waiting for postgres	{"attempt": 1}
    2023-03-31T18:14:00.229+0200	INFO	waiting for postgres	{"attempt": 2}
    2023-03-31T18:14:01.311+0200	INFO	waiting for postgres	{"attempt": 3}
    2023-03-31T18:14:02.623+0200	INFO	waiting for postgres	{"attempt": 4}
    2023-03-31T18:14:04.207+0200	INFO	waiting for postgres	{"attempt": 5}
    2023-03-31T18:14:06.550+0200	INFO	waiting for postgres	{"attempt": 6}
    2023-03-31T18:14:11.062+0200	INFO	waiting for postgres	{"attempt": 7}

**Review by adriantam — APPROVED** [review]

**Review by adriantam — DISMISSED** [review]

File: cmd/migrate_test.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+	confdir := homedir + string(os.PathSeparator) + ".openfga"
+	require.Nil(t, os.Mkdir(confdir, 0750))
+
+	return confdir
+}
+
+func prepareTempConfigFile(t *testing.T, config string) {
+	confdir := prepareTempConfigDir(t)
+	confFile, err := os.Create(confdir + string(os.PathSeparator) + "config.yaml")
```
Comment (miparnisari) [inline cmd/migrate_test.go:30]:
```suggestion
	os.Create(filepath.Join(confdir, "config.yaml"))
```

File: cmd/migrate_test.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+const defaultDuration = 1 * time.Minute
+
+func prepareTempConfigDir(t *testing.T) string {
+	_, err := os.Stat("/etc/openfga/config.yaml")
+	require.ErrorIs(t, err, os.ErrNotExist, "Config file at /etc/openfga/config.yaml would disturb test result.")
+
+	homedir := t.TempDir()
+	t.Setenv("HOME", homedir)
+
+	confdir := homedir + string(os.PathSeparator) + ".openfga"
```
Comment (miparnisari) [inline cmd/migrate_test.go:22]:
```suggestion
	confdir := filepath.Join(homedir, ".openfga")
```

**Review by miparnisari — DISMISSED** [review]

**Review by miparnisari — APPROVED** [review]

**Pull request merged**